<a href="https://colab.research.google.com/github/chrisjallaine/CloudComputingDataEngineeringActivity/blob/main/ETL_Phase_Data_Warehouse.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Dependencies

In [ ]:
import pandas as pd
import sqlalchemy
import numpy as np
from scipy import stats
from sqlalchemy import create_engine

## Load the Data

In [ ]:
# Connect to database
conn = sqlalchemy.create_engine('postgresql://user:U02rC3OwlPV8oEZO39S43P7w3uaLHLA1@dpg-d0pbji8dl3ps73akk2i0-a.singapore-postgres.render.com/rawsales')

In [160]:
# Read the data from the database into 'data' DataFrame
data = pd.read_sql('SELECT * FROM hardware_sales', conn)

In [161]:
data.head()

sls_ord_num     prd_key    CID  sls_order_dt  sls_ship_dt  sls_due_dt  \
0     SO43697  BK-R93R-62  21768      20101229     20110105    20110110   
1     SO43698  BK-M82S-44  28389      20101229     20110105    20110110   
2     SO43699  BK-M82S-44  25863      20101229     20110105    20110110   
3     SO43700  BK-R50B-62  14501      20101229     20110105    20110110   
4     SO43700  BK-R50B-62  14501      20101229     20110105    20110110   

   sls_sales  sls_quantity  sls_price       BDATE  ... prd_id  \
0     3578.0             1     3578.0  1952-02-19  ...    310   
1     3400.0             1     3400.0  1970-06-17  ...    346   
2     3400.0             1     3400.0  1952-06-01  ...    346   
3      699.0             1      699.0  1943-11-10  ...    336   
4      699.0             1      699.0  1943-11-10  ...    337   

                    prd_nm prd_cost prd_line prd_start_dt  prd_end_dt     ID  \
0         Road-150 Red- 62   2171.0       R    2011-07-01  2007-12-28  BI_RB   
1  Mountain-100 Silver- 44   1912.0       M    2011-07-01  2007-12-28  BI_MB   
2  Mountain-100 Silver- 44   1912.0       M    2011-07-01  2007-12-28  BI_MB   
3       Road-650 Black- 62    413.0       R    2011-07-01  2007-12-28  BI_RB   
4       Road-650 Black- 62    487.0       R    2012-07-01  2008-12-27  BI_RB   

     CAT          SUBCAT MAINTENANCE  
0  Bikes      Road Bikes         Yes  
1  Bikes  Mountain Bikes         Yes  
2  Bikes  Mountain Bikes         Yes  
3  Bikes      Road Bikes         Yes  
4  Bikes      Road Bikes         Yes  

[5 rows x 28 columns]

## Data Information

In [162]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 89833 entries, 0 to 89832
Data columns (total 28 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   sls_ord_num         89833 non-null  object 
 1   prd_key             89833 non-null  object 
 2   CID                 89833 non-null  int64  
 3   sls_order_dt        89833 non-null  int64  
 4   sls_ship_dt         89833 non-null  int64  
 5   sls_due_dt          89833 non-null  int64  
 6   sls_sales           89818 non-null  float64
 7   sls_quantity        89833 non-null  int64  
 8   sls_price           89822 non-null  float64
 9   BDATE               89833 non-null  object 
 10  GEN                 83450 non-null  object 
 11  cst_id              89833 non-null  float64
 12  cst_firstname       89823 non-null  object 
 13  cst_lastname        89822 non-null  object 
 14  cst_marital_status  89826 non-null  object 
 15  cst_gndr            58500 non-null  object 
 16  cst_

## Metadata

# Dataset Metadata

| Column Name            | Non-Null Count | Data Type | Description (Inferred)                     | Notes                              |
|------------------------|----------------|-----------|--------------------------------------------|------------------------------------|
| **sls_ord_num**        | 89,833         | `object`  | Sales order number (unique identifier)     | Likely alphanumeric.               |
| **prd_key**           | 89,833         | `object`  | Product key (identifier)                   | Foreign key for products.          |
| **CID**               | 89,833         | `int64`   | Customer ID                                | Unique customer identifier.        |
| **sls_order_dt**      | 89,833         | `int64`   | Order date (numeric format)                | Needs conversion to datetime.      |
| **sls_ship_dt**       | 89,833         | `int64`   | Shipment date (numeric format)             | Needs conversion to datetime.      |
| **sls_due_dt**        | 89,833         | `int64`   | Due date (numeric format)                  | Needs conversion to datetime.      |
| **sls_sales**         | 89,818         | `float64` | Total sales amount                         | 15 missing values.                 |
| **sls_quantity**      | 89,833         | `int64`   | Quantity of products sold                  |                                    |
| **sls_price**         | 89,822         | `float64` | Unit price of product                      | 11 missing values.                 |
| **BDATE**             | 89,833         | `object`  | Birthdate (string format)                  | Needs datetime parsing.            |
| **GEN**               | 83,450         | `object`  | Generation (e.g., Gen X, Millennial)       | 6,383 missing values.              |
| **cst_id**            | 89,833         | `float64` | Customer ID (alternative)                  | Duplicate of `CID`?                |
| **cst_firstname**     | 89,823         | `object`  | Customer first name                        | 10 missing values.                 |
| **cst_lastname**      | 89,822         | `object`  | Customer last name                         | 11 missing values.                 |
| **cst_marital_status**| 89,826         | `object`  | Marital status (e.g., Single, Married)     | 7 missing values.                  |
| **cst_gndr**          | 58,500         | `object`  | Customer gender                            | 31,333 missing values (35%).       |
| **cst_create_date**   | 89,833         | `object`  | Customer account creation date             | Needs datetime parsing.            |
| **CNTRY**             | 88,484         | `object`  | Customer country                           | 1,349 missing values.              |
| **prd_id**            | 89,833         | `int64`   | Product ID                                 | Unique identifier.                 |
| **prd_nm**            | 89,833         | `object`  | Product name                               |                                    |
| **prd_cost**          | 89,833         | `float64` | Product cost                               |                                    |
| **prd_line**          | 89,833         | `object`  | Product line/category                      | e.g., Electronics, Apparel.        |
| **prd_start_dt**      | 89,833         | `object`  | Product launch date                        | Needs datetime parsing.            |
| **prd_end_dt**        | 32,667         | `object`  | Product discontinuation date               | 57,166 missing (likely active).    |
| **ID**                | 89,833         | `object`  | Additional identifier (ambiguous)          | Check business context.            |
| **CAT**               | 89,833         | `object`  | Product category                           | e.g., Hardware, Software.          |
| **SUBCAT**            | 89,833         | `object`  | Product subcategory                        |                                    |
| **MAINTENANCE**       | 89,833         | `object`  | Maintenance flag (e.g., Yes/No)            | Binary or descriptive?             |

### Key Observations:
1. **Temporal Columns**:  
   - Dates (`sls_order_dt`, `sls_ship_dt`, `BDATE`, etc.) are stored as `int64` or `object`—need conversion to datetime.
2. **Missing Values**:  
   - `cst_gndr` (35% missing), `prd_end_dt` (64% missing), `GEN` (7% missing).
3. **Potential Duplicates**:  
   - `CID` (int) vs. `cst_id` (float)—verify if they represent the same entity.
4. **Product Lifecycle**:  
   - `prd_end_dt` missing for most entries, suggesting active products.
5. **Sales Metrics**:  
   - `sls_sales`, `sls_price` have minimal missing values (<0.02%).

## Data Cleaning

### Convert Date Columns

- Columns stored as int64 or object need conversion to datetime

In [163]:
#  Fix and rename the messy date columns
def fix_messy_date(val):
    try:
        val_str = str(int(val))
        if len(val_str) == 8:
            return pd.to_datetime(val_str, format='%Y%m%d', errors='coerce').date()
    except:
        return pd.NaT
    return pd.NaT

messy_date_map = {
    'sls_order_dt': 'order_date',
    'sls_ship_dt': 'ship_date',
    'sls_due_dt': 'due_date'
}

for old_col, new_col in messy_date_map.items():
    if old_col in data.columns:
        data[new_col] = data[old_col].apply(fix_messy_date)
        data.drop(old_col, axis=1, inplace=True)
    else:
        print(f"Column '{old_col}' not found — skipping.")

#  Clean conversion for already well-formed date columns
clean_columns_map = {
    'BDATE': 'birth_date',
    'cst_create_date': 'customer_creation_date',
    'prd_start_dt': 'product_start_date',
    'prd_end_dt': 'product_end_date'
}

for old_col, new_col in clean_columns_map.items():
    if old_col in data.columns:
        data[new_col] = pd.to_datetime(data[old_col], errors='coerce').dt.date
        data.drop(old_col, axis=1, inplace=True)
    else:
        print(f"Column '{old_col}' not found — skipping.")

In [164]:
data.head()

sls_ord_num     prd_key    CID  sls_sales  sls_quantity  sls_price     GEN  \
0     SO43697  BK-R93R-62  21768     3578.0             1     3578.0    None   
1     SO43698  BK-M82S-44  28389     3400.0             1     3400.0  Female   
2     SO43699  BK-M82S-44  25863     3400.0             1     3400.0  Female   
3     SO43700  BK-R50B-62  14501      699.0             1      699.0    Male   
4     SO43700  BK-R50B-62  14501      699.0             1      699.0    Male   

    cst_id cst_firstname cst_lastname  ...    CAT          SUBCAT MAINTENANCE  \
0  21768.0          Cole       Watson  ...  Bikes      Road Bikes         Yes   
1  28389.0       Rachael     Martinez  ...  Bikes  Mountain Bikes         Yes   
2  25863.0        Sydney       Wright  ...  Bikes  Mountain Bikes         Yes   
3  14501.0         Ruben       Prasad  ...  Bikes      Road Bikes         Yes   
4  14501.0         Ruben       Prasad  ...  Bikes      Road Bikes         Yes   

   order_date   ship_date    due_date  birth_date customer_creation_date  \
0  2010-12-29  2011-01-05  2011-01-10  1952-02-19             2026-01-05   
1  2010-12-29  2011-01-05  2011-01-10  1970-06-17             2026-01-25   
2  2010-12-29  2011-01-05  2011-01-10  1952-06-01             2026-01-14   
3  2010-12-29  2011-01-05  2011-01-10  1943-11-10             2025-10-12   
4  2010-12-29  2011-01-05  2011-01-10  1943-11-10             2025-10-12   

  product_start_date product_end_date  
0         2011-07-01       2007-12-28  
1         2011-07-01       2007-12-28  
2         2011-07-01       2007-12-28  
3         2011-07-01       2007-12-28  
4         2012-07-01       2008-12-27  

[5 rows x 28 columns]

## Handle Missing Values

### High-Missing Columns:
*cst_gndr* (35% missing):

- Impute with "Unknown" or use predictive modeling if gender is critical.

*prd_end_dt* (64% missing):

- Missing likely indicates active products. Create a flag.

In [165]:
data['is_active_product'] = data['product_end_date'].isna().astype(int)

In [166]:
data['is_active_product'] = data['product_end_date'].isna().astype(int)

## Low-Missing Columns

*sls_sales, sls_price, cst_firstname*, etc.:

- Drop rows or impute with median/mean (numeric) or mode (categorical):



In [167]:
numeric_cols = ['sls_sales', 'sls_price', 'prd_cost']
data[numeric_cols] = data[numeric_cols].fillna(data[numeric_cols].median())

categorical_cols = ['cst_firstname', 'cst_lastname', 'cst_marital_status']
data[categorical_cols] = data[categorical_cols].fillna('Not Specified')

## Standardize Categorical Data

*GEN* (Generation), *CNTRY* (Country), *prd_line*:

- Convert to uppercase/lowercase and trim whitespace

In [168]:
data['GEN'] = data['GEN'].str.upper().str.strip()
data['CNTRY'] = data['CNTRY'].str.upper().str.strip()

*cst_gndr*

- Standardize values (e.g., "M", "F", "Unknown")

In [169]:
gender_map = {'Male': 'M', 'Female': 'F', 'm': 'M', 'f': 'F'}
data['cst_gndr'] = data['cst_gndr'].replace(gender_map).fillna('Unknown')

## Deduplicate and Validate Identifiers

Check *CID* vs. *cst_id*

In [170]:
if data['CID'].equals(data['cst_id'].astype('int64')):
    data.drop('cst_id', axis=1, inplace=True)  # Remove duplicate

## Validate sls_ord_num (Sales Order Number)

- Ensure uniqueness

In [172]:
duplicate_count = len(data) - data['sls_ord_num'].nunique()
print(f"{duplicate_count} duplicate order numbers found.")

62174 duplicate order numbers found.


In [173]:
data = data.drop_duplicates(subset='sls_ord_num', keep='first')

## Renaming

In [174]:
data['cst_marital_status'].unique()

array(['S', 'M', 'Not Specified'], dtype=object)

In [175]:
data['cst_marital_status'] = data['cst_marital_status'].replace({
    'S': 'Single',
    'M': 'Married'
})

## Handle Product Data
*Active* vs. *Discontinued* Products:

- Use product_end_date to flag status.

### Product Categories (CAT, SUBCAT):

- Merge or split if hierarchies are unclear:

In [176]:
data['full_category'] = data['CAT'] + ' > ' + data['SUBCAT']

## Fix Numeric Outliers
*sls_quantity*, *sls_price*:

- Clip extreme values (e.g., beyond 3 standard deviations)

In [177]:
data = data[(np.abs(stats.zscore(data[['sls_quantity', 'sls_price']])) < 3).all(axis=1)]

## Feature Engineering

In [178]:
# Customer Age

# Convert birth_date to pandas datetime (Timestamp)
data['birth_date'] = pd.to_datetime(data['birth_date'], errors='coerce')

# Calculate age in years (floor division)
data['customer_age'] = (pd.to_datetime('today') - data['birth_date']).dt.days // 365

In [179]:
data['customer_age'] = data['customer_age'].astype('Int64')

In [180]:
# Order-to-Ship Delay
print(data[['order_date', 'ship_date']].dtypes)

order_date    object
ship_date     object
dtype: object


In [181]:
for col in ['order_date', 'ship_date']:
    data[col] = pd.to_datetime(data[col], errors='coerce')

In [182]:
data['order_to_ship_days'] = (data['ship_date'] - data['order_date']).dt.days

## Drop Unecessary Columns

In [183]:
data.columns

Index(['sls_ord_num', 'prd_key', 'CID', 'sls_sales', 'sls_quantity',
       'sls_price', 'GEN', 'cst_firstname', 'cst_lastname',
       'cst_marital_status', 'cst_gndr', 'CNTRY', 'prd_id', 'prd_nm',
       'prd_cost', 'prd_line', 'ID', 'CAT', 'SUBCAT', 'MAINTENANCE',
       'order_date', 'ship_date', 'due_date', 'birth_date',
       'customer_creation_date', 'product_start_date', 'product_end_date',
       'is_active_product', 'full_category', 'customer_age',
       'order_to_ship_days'],
      dtype='object')

In [184]:
data.drop(columns=[
    'full_category', 'cst_gndr'
], inplace=True)

## Validation

data.drop(columns=[
    'cst_gender',
    'ship_date_time',
    'due_date_time',
    'full_category'
], inplace=True)

In [185]:
data.isna().sum()

sls_ord_num                   0
prd_key                       0
CID                           0
sls_sales                     0
sls_quantity                  0
sls_price                     0
GEN                        1820
cst_firstname                 0
cst_lastname                  0
cst_marital_status            0
CNTRY                       332
prd_id                        0
prd_nm                        0
prd_cost                      0
prd_line                      0
ID                            0
CAT                           0
SUBCAT                        0
MAINTENANCE                   0
order_date                    8
ship_date                     0
due_date                      0
birth_date                    9
customer_creation_date        0
product_start_date            0
product_end_date          17088
is_active_product             0
customer_age                  9
order_to_ship_days            8
dtype: int64

In [186]:
data = data.dropna()

In [187]:
data['order_to_ship_days'] = data['order_to_ship_days'].astype('Int64')

In [188]:
for col in ['due_date', 'customer_creation_date', 'product_start_date', 'product_end_date']:
    data[col] = pd.to_datetime(data[col], errors='coerce').dt.date

In [189]:
data['is_active_product'] = data['is_active_product'].astype(bool)

In [190]:
for col in ['order_date', 'ship_date', 'due_date']:
    data[col] = pd.to_datetime(data[col], errors='coerce').dt.date


In [193]:
date_cols = [
    'order_date', 'ship_date', 'due_date',
    'birth_date', 'customer_creation_date',
    'product_start_date', 'product_end_date'
]

for col in date_cols:
    data[col] = pd.to_datetime(data[col], errors='coerce')

In [194]:
for col in date_cols:
    data[f'{col}_month'] = data[col].dt.month_name()

In [195]:
data.head()

sls_ord_num     prd_key    CID  sls_sales  sls_quantity  sls_price     GEN  \
1     SO43698  BK-M82S-44  28389     3400.0             1     3400.0  FEMALE   
2     SO43699  BK-M82S-44  25863     3400.0             1     3400.0  FEMALE   
3     SO43700  BK-R50B-62  14501      699.0             1      699.0    MALE   
5     SO43701  BK-M82S-44  11003     3400.0             1     3400.0  FEMALE   
6     SO43702  BK-R93R-44  27645     3578.0             1     3578.0    MALE   

  cst_firstname cst_lastname cst_marital_status  ... is_active_product  \
1       Rachael     Martinez             Single  ...             False   
2        Sydney       Wright             Single  ...             False   
3         Ruben       Prasad            Married  ...             False   
5       Christy          Zhu             Single  ...             False   
6         Colin        Anand            Married  ...             False   

   customer_age order_to_ship_days  order_date_month ship_date_month  \
1            54                  7          December         January   
2            73                  7          December         January   
3            81                  7          December         January   
5            51                  7          December         January   
6            52                  7          December         January   

  due_date_month birth_date_month customer_creation_date_month  \
1        January             June                      January   
2        January             June                      January   
3        January         November                      October   
5        January           August                      October   
6        January           August                      January   

  product_start_date_month product_end_date_month  
1                     July               December  
2                     July               December  
3                     July               December  
5                     July               December  
6                     July               December  

[5 rows x 36 columns]

In [196]:
# Export to CSV without the DataFrame index (this is jsut for checking)
data.to_csv('dataset.csv', index=False)

## Sending to Data Warehouse

In [ ]:
!pip install pandas sqlalchemy psycopg2-binary

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.0/3.0 MB 24.1 MB/s eta 0:00:00


In [197]:
#Render database details
DATABASE_URL = "postgresql://user:szrVeYFcCzLxNRLlxxPtX6w4fWqM1ulr@dpg-d0oun3muk2gs73945erg-a.singapore-postgres.render.com/salesdashboard_of09"

In [199]:
from sqlalchemy import create_engine

# Create a SQLAlchemy engine
engine = create_engine(DATABASE_URL)

# Export DataFrame to PostgreSQL
data.to_sql(
    name='dashboard_data',
    con=engine,
    if_exists='replace',
    index=False,
    chunksize=1000
)

print("Data successfully exported to Render PostgreSQL!")

Data successfully exported to Render PostgreSQL!


In [200]:
# Read back data to confirm
df_loaded = pd.read_sql('SELECT * FROM dashboard_data LIMIT 10', con=engine)
df_loaded

sls_ord_num     prd_key    CID  sls_sales  sls_quantity  sls_price     GEN  \
0     SO43698  BK-M82S-44  28389     3400.0             1     3400.0  FEMALE   
1     SO43699  BK-M82S-44  25863     3400.0             1     3400.0  FEMALE   
2     SO43700  BK-R50B-62  14501      699.0             1      699.0    MALE   
3     SO43701  BK-M82S-44  11003     3400.0             1     3400.0  FEMALE   
4     SO43702  BK-R93R-44  27645     3578.0             1     3578.0    MALE   
5     SO43703  BK-R93R-62  16624     3578.0             1     3578.0    MALE   
6     SO43704  BK-M82B-48  11005     3375.0             1     3375.0    MALE   
7     SO43705  BK-M82S-38  11011     3400.0             1     3400.0    MALE   
8     SO43706  BK-R93R-48  27621     3578.0             1     3578.0    MALE   
9     SO43707  BK-R93R-48  27616     3578.0             1     3578.0  FEMALE   

  cst_firstname cst_lastname cst_marital_status  ... is_active_product  \
0       Rachael     Martinez             Single  ...             False   
1        Sydney       Wright             Single  ...             False   
2         Ruben       Prasad            Married  ...             False   
3       Christy          Zhu             Single  ...             False   
4         Colin        Anand            Married  ...             False   
5        Albert      Alvarez             Single  ...             False   
6         Julio         Ruiz             Single  ...             False   
7        Curtis           Lu            Married  ...             False   
8        Edward        Brown             Single  ...             False   
9          Emma        Brown             Single  ...             False   

   customer_age order_to_ship_days  order_date_month ship_date_month  \
0            54                  7          December         January   
1            73                  7          December         January   
2            81                  7          December         January   
3            51                  7          December         January   
4            52                  7          December         January   
5            41                  7          December         January   
6            48                  7          December         January   
7            56                  7          December         January   
8            50                  7          December         January   
9            53                  7          December         January   

  due_date_month birth_date_month customer_creation_date_month  \
0        January             June                      January   
1        January             June                      January   
2        January         November                      October   
3        January           August                      October   
4        January           August                      January   
5        January             July                      January   
6        January           August                      October   
7        January              May                      October   
8        January          January                      January   
9        January             June                      January   

  product_start_date_month product_end_date_month  
0                     July               December  
1                     July               December  
2                     July               December  
3                     July               December  
4                     July               December  
5                     July               December  
6                     July               December  
7                     July               December  
8                     July               December  
9                     July               December  

[10 rows x 36 columns]

---